### PR TITLE
BUG: Fix segfault in `random.permutation(x)` when  x is a string.

### PR DESCRIPTION
**Pull request by maxwell-aladago**
Reference  [ #14238](https://github.com/numpy/numpy/issues/14238)

I made the assumption that anyone who passes a string  `x` to the function `np.random.permutation(x)` most likely intend to shuffle the string in question. I suggest that if this isn't the desired behaviour, then I can flag an error instead of the segfault.

### CODE REVIEW

**Comment by seberg** [comment]
To be honest, I do not think I like to special case strings like that. This also does not fix the undrelying segmentation fault, since the same error will still happen for example for floating point input.

The underlying issue is the cython `boundscheck=False` directive, which also applies to tuples. Either, we have to mark this (and possibly some other places) expliciteley using:
```
with cython.boundscheck(True):
    ...
```
or the other way around (could probably also mark the whole function).

Maybe the best option here specifically is to just raise a new `ValueError("permutation of a 0 dimensional array is not defined.")` or something nicer.

**Comment by rkern** [comment]
The old behavior is to raise an `IndexError`. We should probably do just that for `RandomState.permutation()`. We can think about what we want for `Generator.permutation('some_string')`, but we should have the `IndexError` as a backstop for `Generator.permutation(1.5)`, etc.

```
[~]
|1> np.random.permutation('abc')
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-1-eab1612dcd44> in <module>()
----> 1 np.random.permutation('abc')

mtrand.pyx in mtrand.RandomState.permutation()

IndexError: tuple index out of range
```

**Comment by seberg** [comment]
Hmm, right, at least for the legacy API its probably better to err on the save side. Another option, which works for both is to use `np.AxisError` (since effectively there is an `axis=0` going on, even if  the user cannot set it). That error subclasses from IndexError and ValueError for exactly that backward compatibility concern.

**Comment by maxwell-aladago** [comment]
Thanks for the feedback.  The issue about floating point numbers is true. Will you consider a PR for an np.AxisError or you'll prefer an explicit cython bounds check? @seberg

**Comment by seberg** [comment]
Raising an error explicitly is good, I do not mind much either way, but I feel AxisError may be nicer than IndexError.
If you have time, we may want to check the file for similar issues (e.g. bad Axis input that used to raise an IndexError automatically, and now will segfault).

**Comment by maxwell-aladago** [comment]
> Raising an error explicitly is good, I do not mind much either way, but I feel AxisError may be nicer than IndexError.
> If you have time, we may want to check the file for similar issues (e.g. bad Axis input that used to raise an IndexError automatically, and now will segfault).

I do have some time. Will look at the bad Axis input too.

**Comment by eric-wieser** [comment]
My claim would be that AxisError is only correct if you can trigger it via a sane call to normalize_axis_index

**Comment by maxwell-aladago** [comment]
> My claim would be that AxisError is only correct if you can trigger it via a sane call to normalize_axis_index

 A check like 

'.normalize_axis_index(0, arr.ndim)` will trigger an AxisError! And I think it's sane enough since intuitively, we expect all permutable inputs to be either ints or have at least 1 axis

File: numpy/random/generator.pyx [hunk 1]
```diff
@@ -3921,7 +3921,7 @@ cdef class Generator:
         Randomly permute a sequence, or return a permuted range.
 
         If `x` is a multi-dimensional array, it is only shuffled along its
-        first index.
+        first index.
```
Comment (mattip) [inline numpy/random/generator.pyx]: Please revert the whitespace additions

File: numpy/random/tests/test_generator_mt19937.py [hunk 2]
```diff
@@ -758,6 +758,12 @@ def test_permutation(self):
         actual = random.permutation(arr_2d)
         assert_array_equal(actual, np.atleast_2d(desired).T)
 
+        bad_x_str = "abcd"
+        assert_raises(np.AxisError, random.permutation, bad_x_str)
+
+        bad_x_float = 1.2
+        assert_raises(np.AxisError, random.permutation, bad_x_float)
+
```
Comment (mattip) [inline numpy/random/tests/test_generator_mt19937.py]: This block and the try, except block above should be the only changes in this PR. All the other formatting changes should be reverted.

**Comment by anntzer** [comment]
fwiw I think this should raise an error: this would be consistent with np.random.choice("abcd") which currently raises an error -- in fact, this would even be consistent with np.asarray("abcd") returning a scalar array of dtype <U4 rather than a shape-(4,) array.

**Comment by maxwell-aladago** [comment]
> fwiw I think this should raise an error: this would be consistent with np.random.choice("abcd") which currently raises an error -- in fact, this would even be consistent with np.asarray("abcd") returning a scalar array of dtype <U4 rather than a shape-(4,) array.

`np.random.choice("abc")` currently gives a `ValueError` whereas
`np.random.permutation("abc")` results in a `segfault`. The current patch ensures it throws a sensible `np.AxisError`. To be consistent with `np.random.choice("abc")`, we'll have to update the patch to throw a `ValueError` instead.

**Comment by WarrenWeckesser** [comment]
Wouldn't a `TypeError` or `ValueError` be more appropriate?  The argument for raising `AxisError` is not very strong, whereas the text added to the exception message, `"x must be an integer or an array_like object."`, suggests that the exception should be a `TypeError` or a `ValueError`.

**Comment by maxwell-aladago** [comment]
> Wouldn't a `TypeError` or `ValueError` be more appropriate? The argument for raising `AxisError` is not very strong, whereas the text added to the exception message, `"x must be an integer or an array_like object."`, suggests that the exception should be a `TypeError` or a `ValueError`.

Changed it to `ValueError`

**Comment by eric-wieser** [comment]
`ValueError` is fine. If we added an explicit `axis` argument, then we can switch to AxisError then, especially since `issubclass(AxisError, ValueError)`.

**Comment by rkern** [comment]
The appropriate comparison is not `np.random.choice()` but the 1.16 behavior of `np.random.permutation()`, which, as noted above, raises an `IndexError`.

**Comment by WarrenWeckesser** [comment]
So, for the legacy code, we should raise an `IndexError` for backwards compatbility:
```
In [2]: np.__version__                                                          
Out[2]: '1.16.4'

In [3]: np.random.permutation("abc")                                            
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-3-a4a90726d5df> in <module>
----> 1 np.random.permutation("abc")

mtrand.pyx in mtrand.RandomState.permutation()

IndexError: tuple index out of range
```

From a user's point of view, however, that exception is not very helpful.  Why is it an *Index*Error? What is the "tuple index" referred to in the error message?  I don't see why we would design the API to raise that exception, and I don't think we should maintain that behavior in the new `Generator` code.  What do you think of a `ValueError` for `Generator.permutation` in this case?

**Comment by WarrenWeckesser** [comment]
Another question about fixing this issue.   To what extent do we expect a scalar array (i.e. a zero-dimensional array) to act like an instance of its underlying dtype?  Currently, the seg. fault occurs when a zero-dimensional array is passed to `permutation`:

```
In [2]: import numpy as np                                                                                                         

In [3]: rng = np.random.Generator(np.random.PCG64())                                                                               

In [4]: rng.permutation(5)                                                                                                         
Out[4]: array([0, 3, 4, 1, 2])

In [5]: rng.permutation(np.int64(5))                                                                                               
Out[5]: array([2, 1, 4, 3, 0])

In [6]: rng.permutation(np.array(5))                                                                                               
Segmentation fault: 11
```
Should that last case act the same as the previous two, and return a permutation of `range(5)`?

**Comment by seberg** [comment]
I think the last case we could just stick with raising an error, but either will work. While we do generally allow 0D arrays instead of scalars, the function is overloaded in a way that I think it would be fair to assume it is unintentional when it happens. Plus 0D arrays are exceedingly rare in any case.

**Comment by maxwell-aladago** [comment]
> Another question about fixing this issue. To what extent do we expect a scalar array (i.e. a zero-dimensional array) to act like an instance of its underlying dtype? Currently, the seg. fault occurs when a zero-dimensional array is passed to `permutation`:
> 
> ```
> In [2]: import numpy as np                                                                                                         
> 
> In [3]: rng = np.random.Generator(np.random.PCG64())                                                                               
> 
> In [4]: rng.permutation(5)                                                                                                         
> Out[4]: array([0, 3, 4, 1, 2])
> 
> In [5]: rng.permutation(np.int64(5))                                                                                               
> Out[5]: array([2, 1, 4, 3, 0])
> 
> In [6]: rng.permutation(np.array(5))                                                                                               
> Segmentation fault: 11
> ```
> 
> Should that last case act the same as the previous two, and return a permutation of `range(5)`?

my patch actually fixed the case for '`np.random.permutation(np.array(5))`

**Comment by eric-wieser** [comment]
I don't think we should use `IndexError` for the new Generator objects - part of the point of the new API was to allow us to break compatibility. ValueError was in my opinion more appropriate.

**Comment by maxwell-aladago** [comment]
> I don't think we should use `IndexError` for the new Generator objects - part of the point of the new API was to allow us to break compatibility. ValueError was in my opinion more appropriate.

So, you want me to roll it back to ValueError? @eric-wieser. ValueError looks right to me too because it really isn't an indexing problem.

**Comment by rkern** [comment]
Ah, per the PR title, I thought we were talking about `np.random.permutation()` which is `RandomState.permutation()`. Do what you like with `Generator.permutation()`. `AxisError` would be the best exception there (and `Generator.choice()` to match) in the cases that you don't want to handle specially.

But it seems like this PR does not fix `RandomState.permutation()`, so `np.random.permutation()` remains unfixed as well. That still segfaults. The proper fix there is to raise an `IndexError` to match 1.16 behavior.

**Comment by maxwell-aladago** [comment]
> Ah, per the PR title, I thought we were talking about `np.random.permutation()` which is `RandomState.permutation()`. Do what you like with `Generator.permutation()`. `AxisError` would be the best exception there (and `Generator.choice()` to match) in the cases that you don't want to handle specially.
> 
> But it seems like this PR does not fix `RandomState.permutation()`, so `np.random.permutation()` remains unfixed as well. That still segfaults. The proper fix there is to raise an `IndexError` to match 1.16 behavior.

Fixes both `RandomState.permutatio()` and `Generator.permutation()` now.  Let me know whether this is a good fix @rkern so I can change change the `Generator.choice()` to match too.

**Comment by WarrenWeckesser** [comment]
@rkern wrote
> Do what you like with Generator.permutation(). AxisError would be the best exception there (and Generator.choice() to match) in the cases that you don't want to handle specially.

What is the rationale for raising an `AxisError` instead of a `ValueError`?  For an input such as `x=1.5`, is it something like *"The argument must be int or array_like.  If the argument is not, in fact, an int, then it must be array_like, and when we try to use the first axis of the scalar 1.5 we get an `AxisError`"*?

File: numpy/random/generator.pyx (hunk 1 above)
Comment (seberg) [reply numpy/random/generator.pyx]: The spurious space is still there.

File: numpy/random/generator.pyx [hunk 3]
```diff
@@ -3935,6 +3935,10 @@ cdef class Generator:
         out : ndarray
             Permuted sequence or array range.
 
+        Raises
+        ------
+        np.AxisError: when x is a  0-dimensional array which is not an integer
```
Comment (seberg) [inline numpy/random/generator.pyx]: I do not think we put this for errors that are due to bad input?

File: numpy/random/generator.pyx [hunk 4]
```diff
@@ -3950,13 +3954,19 @@ cdef class Generator:
                [0, 1, 2],
                [3, 4, 5]])
 
+        >>> rng.permutation("abc")
+        Traceback (most recent call last):
+            ...
+        numpy.AxisError: x must be an integer or at least 1-dimensional
```
Comment (seberg) [inline numpy/random/generator.pyx]: Not sure an error example helps too much here, but maybe if someone thinks they can shuffle strings or similar objects.

File: numpy/random/generator.pyx [hunk 5]
```diff
@@ -3950,13 +3954,19 @@ cdef class Generator:
...
         """
-        if isinstance(x, (int, np.integer)):
+        arr = np.asarray(x)
+        if arr.ndim < 1 and arr.dtype.kind == 'i':
```
Comment (seberg) [inline numpy/random/generator.pyx]: I do not mind raising the error personally for `np.array(1)`. The old check seems OK to me?

File: numpy/random/mtrand.pyx [hunk 6]
```diff
@@ -4149,12 +4153,14 @@ cdef class RandomState:
                [3, 4, 5]])
 
         """
-        if isinstance(x, (int, np.integer)):
+        arr = np.asarray(x)
+        if arr.ndim < 1 and arr.dtype.kind == 'i':
```
Comment (seberg) [inline numpy/random/mtrand.pyx]: If we guard the old API well, I guess we should not change it here in either case?

**Comment by rkern** [comment]
> What is the rationale for raising an `AxisError` instead of a `ValueError`?

Broadly speaking, we are expecting to be indexing along the first of >=1 axes, in the usual case. The integer semantic branch complicates that, and the message should incorporate information about that, but I think that `AxisError` is the most reasonable (as it is both a `ValueError` and an `IndexError`).

**Comment by charris** [comment]
The refguide check is failing.

**Comment by charris** [comment]
Everyone good with this?

**Comment by charris** [comment]
@maxwell-aladago Could you make another PR against 1.16.x that only makes the fixes/tests for RandomState?

File: numpy/random/generator.pyx (hunk 1 above)
Comment (WarrenWeckesser) [reply numpy/random/generator.pyx]: Still there.  @maxwell-aladago, the space added at the end of this line should not be there.

**Comment by maxwell-aladago** [comment]
> @maxwell-aladago Could you make another PR against 1.16.x that only makes the fixes/tests for RandomState?

I am trying to but the PR can't be merged with master. @charris , Can you advise me how to go about this? Thank you

**Comment by rkern** [comment]
I don't think there's a segfault in 1.16. It was introduced in 1.17.

**Comment by charris** [comment]
> I don't think there's a segfault in 1.16. It was introduced in 1.17.

Ah, just checked and you are right.

**Comment by seberg** [comment]
Did anyone check for other similar errors, or should I go through it?

**Pull request merged**